packages/core/src/external/carequality/ihe-gateway-v2/inbound/shared.ts

```diff
@@ -0,0 +1,131 @@
+};
+
+export function convertSamlHeaderToAttributes(header: SamlHeader): SamlAttributes {
+  const attributes = header.Security.Assertion.AttributeStatement.Attribute;
+
+  const getAttributeValue = (name: string): string | undefined => {
+    const attribute = attributes.find(attr => attr._Name === name);
+    if (!attribute) return undefined;
+    if (typeof attribute.AttributeValue === "string") return attribute.AttributeValue;
+    if (istextSchema(attribute.AttributeValue)) return extractText(attribute.AttributeValue);
+    return undefined;
+  };
+
+  const getRoleAttributeValue = (name: string): { code: string; display: string } | undefined => {
+    const attribute = attributes.find(attr => attr._Name === name);
+    if (!attribute) return undefined;
+    if (isRoleObject(attribute.AttributeValue)) {
+      return {
+        code: attribute.AttributeValue.Role._code,
+        display: attribute.AttributeValue.Role._displayName,
+      };
+    }
+    return undefined;
+  };
+
+  const getPurposeOfUseAttributeValue = (name: string): string | undefined => {
+    const attribute = attributes.find(attr => attr._Name === name);
+    if (!attribute) return undefined;
+    if (isPurposeOfUseObject(attribute.AttributeValue)) {
+      return attribute.AttributeValue.PurposeOfUse._code;
+    }
+    return undefined;
+  };
+
+  const subjectId = getAttributeValue("urn:oasis:names:tc:xspa:1.0:subject:subject-id");
+  if (!subjectId) {
+    throw new Error("Subject ID is required");
+  }
+
+  const organization = getAttributeValue("urn:oasis:names:tc:xspa:1.0:subject:organization");
+  if (!organization) {
+    throw new Error("Organization is required");
+  }
+
+  const organizationId = getAttributeValue("urn:oasis:names:tc:xspa:1.0:subject:organization-id");
+  if (!organizationId) {
+    throw new Error("Organization ID is required");
+  }
+
+  const homeCommunityId = getAttributeValue("urn:nhin:names:saml:homeCommunityId");
+  if (!homeCommunityId) {
+    throw new Error("Home community ID is required");
+  }
+
+  const subjectRole = getRoleAttributeValue("urn:oasis:names:tc:xacml:2.0:subject:role");
+  if (!subjectRole) {
+    throw new Error("Subject role is required");
+  }
+
+  const purposeOfUse = getPurposeOfUseAttributeValue(
+    "urn:oasis:names:tc:xspa:1.0:subject:purposeofuse"
+  );
+  if (purposeOfUse != treatmentPurposeOfUse) {
+    throw new Error("A treatment purpose of use is required");
+  }
+
+  return {
+    subjectId: subjectId,
+    organization: organization,
+    organizationId: stripUrnPrefix(organizationId),
+    homeCommunityId: stripUrnPrefix(homeCommunityId),
+    subjectRole: subjectRole,
+    purposeOfUse: purposeOfUse,
+  };
+}
+
+export function extractTimestamp(header: SamlHeader): string {
+  return header.Security.Timestamp.Created;
+}
+
+export function createSecurityHeader({
+  signatureConfirmation,
+}: {
+  signatureConfirmation?: string | undefined;
+}) {
```

Comment:
NABD if not, but don't we have a type to represent the return of this function?

Reply:
we dont. I can add the object return type